chip-tool: do not hardcode Thread network id

bluebin14 opened the pull request:
#### Problem
What is being fixed?  Examples:
* chip-tool hardcodes Thread network id to 0123456789abcdef. This does not work in most cases, see issue #8097

#### Change overview
What's in this PR
* use the Thread network id of the dataset instead of the hardcoded network id
* not having the extendedPanId in the dataset is now an error, as it should be, since it is a required item for looking up the dataset (note however that it won't be missing though since code will fail already at the add Thread network command if it is missing).

#### Testing
How was this tested? (at least one bullet point required)
* tested manually using chip-tool and HomePod Mini where Thread network id cannot be hardcoded

andy31415 commented on examples/chip-tool/commands/pairing/PairingCommand.cpp:
```diff
@@ -235,6 +236,21 @@ CHIP_ERROR PairingCommand::EnableNetwork()
+        CHIP_ERROR err = dataset.Init(mOperationalDataset);
+        if (err == CHIP_NO_ERROR)
+        {
+            static uint8_t sExtendedPanId[Thread::kSizeExtendedPanId];
```
usage of static here seems suspicious - seems necessary because lifetime, however could we get away by placing this as a top-level item or making it a member of the PairingCommand instead?

Having a static array somewhat deep inside a method is not ideal.

bluebin14 replied:
Simplified the handling of Thread network id and merged the 2 variables into one. It would be nice if GetExtendedPanId() API gave us a ByteSpan result instead of filling in an external buffer, then we wouldn't even need one.

mspang commented on examples/chip-tool/commands/pairing/PairingCommand.cpp:
```diff
@@ -19,12 +19,14 @@
+constexpr uint64_t kBreadcrumb = 0;
+constexpr uint32_t kTimeoutMs  = 6000;
+
+static uint8_t sThreadNetworkId[] = { 0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef };
```
Please use a local variable.

bluebin14 replied:
Using a local variable would cause undefined behavior because it would go out of scope at this line:
return mCluster.EnableNetwork(successCallback, failureCallback, networkId, kBreadcrumb, kTimeoutMs);
It would need to be a static local variable but that's what I had in the first place until change was requested by @andy31415 to move it to top level or member variable, which I did.

andy31415 replied:
Could it be a pairing command member variable instead?

andy31415 replied:
I see that you updated to use a bytespan in a member object instead - that seems similar (and more compact), at the expense of having the code slightly harder to read (pointer inside data). Marking conversation resolved but leaving up to @mspang to do a review to see if this is acceptable for removing "changes required" flag.

bluebin14 commented:
Rewrote the fix from scratch based on suggestions, removed hardcoded "temporary" id entirely, extendedPanId is a required item, as it should be, print out error if it is missing (it won't be missing though since code will fail already at the add Thread network command already if it is missing).

msandstedt commented on examples/chip-tool/commands/pairing/PairingCommand.cpp:
```diff
@@ -222,6 +221,45 @@ CHIP_ERROR PairingCommand::AddWiFiNetwork()
     return mCluster.AddWiFiNetwork(successCallback, failureCallback, mSSID, mPassword, kBreadcrumb, kTimeoutMs);
 }
 
+chip::ByteSpan PairingCommand::GetThreadNetworkId() const
```
We have a class, `Thread::ThreadOperationalDataset`, for this in `lib/support/ThreadOperationalDataset.h`.

Can that be used to provide this parser?

bluebin14 replied:
that class was exactly what the first 2 versions of this PR used but it had limitations and required extra storage: local storage had the problem of going out of scope when used, global storage was also questioned and indeed unnecessary.

msandstedt replied:
So I can see that `Thread:: ThreadOperationalDataset` doesn't give a zero-copy way to instantiate an instance from a `bytespan`.  But why not just make `mOperationalDataset` of this type in the first place?

Short of that, there still seem to be other reuse opportunities.  For instance, if `mOperationalDataset` can't be of type `Thread:: ThreadOperationalDataset`, why not construct one of these locally?  And if that isn't possible, why can't this parser at least use the `Thread:: ThreadTLV` type?

bluebin14 replied:
Yes we could make mOperationalDataset of type ThreadOperationalDataset but it would only be useful if the API allowed ByteSpan accessors to its internal storage, otherwise we would be back to where we started, needing global or member storage for something that does not logically need it. Note that previous 2 versions of this PR used a local instance of ThreadOperationalDataset and was suffering from same usability issues. Regarding why not at least ThreadTLV type it's very simple: because it is private and cannot be used from the outside.

mspang replied:
I share the concern. ThreadOperationalDataset is our own utility class; we can change it.

bluebin14 replied:
I switched back to using Thread OperationalDataset class as you requested and provided storage for the extendedPanId as a member variable, this way everyone's request can be accommodated.

mspang commented on examples/chip-tool/commands/pairing/PairingCommand.cpp:
```diff
@@ -234,7 +272,7 @@ CHIP_ERROR PairingCommand::EnableNetwork()
     {
-        networkId = ByteSpan(kTemporaryThreadNetworkId, sizeof(kTemporaryThreadNetworkId));
+        networkId = GetThreadNetworkId();
     }
```
Can we handle the case of networkId being empty or at least add a TODO for it?

mspang replied:
Ah sorry you mentioned already why this won't happen.

bluebin14 commented:
Please review new 4th revision which should address everyone's concerns. Note that each revision is equally functional, the differences are only editorial.

Damian-Nordic commented on examples/chip-tool/commands/pairing/PairingCommand.cpp:
```diff
@@ -222,6 +222,32 @@ CHIP_ERROR PairingCommand::AddWiFiNetwork()
+        err = dataset.GetExtendedPanId(mExtendedPanId);
+        if (err == CHIP_NO_ERROR)
+        {
+            threadNetworkId = ByteSpan(mExtendedPanId, sizeof(mExtendedPanId));
```
```suggestion
            return ByteSpan(mExtendedPanId, sizeof(mExtendedPanId));
```

bluebin14 replied:
Did the suggested change.

vivien-apple commented on examples/chip-tool/commands/pairing/PairingCommand.cpp:
```diff
@@ -222,6 +222,32 @@ CHIP_ERROR PairingCommand::AddWiFiNetwork()
+        }
+    }
+
+    if (threadNetworkId.empty())
```
I you reach this code  isn't `threadNetworkId` always empty  ? Maybe you may want to rewrite this code such as: 
```
chip::ByteSpan PairingCommand::GetThreadNetworkId()
{
     // For Thread devices the networkId is the extendedPanId and it is
     // part of the dataset defined by OpenThread

     Thread::OperationalDataset dataset;
     CHIP_ERROR err = dataset.Init(mOperationalDataset);
     if (err == CHIP_NO_ERROR)
     {
         err = dataset.GetExtendedPanId(mExtendedPanId);
         if (err == CHIP_NO_ERROR)
         {
             return ByteSpan(mExtendedPanId, sizeof(mExtendedPanId));
         }
     }

     return ByteSpan();
}
```

bluebin14 replied:
Addressed all reviews but it got autoclosed during force push, had to open new one because Reopen button is disabled.

https://github.com/project-chip/connectedhomeip/pull/8245

vivien-apple commented on examples/chip-tool/commands/pairing/PairingCommand.cpp:
```diff
@@ -234,7 +260,7 @@ CHIP_ERROR PairingCommand::EnableNetwork()
     else
     {
-        networkId = ByteSpan(kTemporaryThreadNetworkId, sizeof(kTemporaryThreadNetworkId));
+        networkId = GetThreadNetworkId();
```
Can we add a check here to check that `networkId` is not empty and returns an error otherwise ? Otherwise the code will continue as if.

bluebin14 replied:
Addressed all reviews but it got autoclosed during force push, had to open new one because Reopen button is disabled.

https://github.com/project-chip/connectedhomeip/pull/8245

franck-apple commented:
@mspang can you help review this?

mspang commented:
> @mspang can you help review this?

Please fix Vivien's comment. Otherwise it looks good now.

bzbarsky-apple commented on examples/chip-tool/commands/pairing/PairingCommand.cpp:
```diff
@@ -222,6 +222,32 @@ CHIP_ERROR PairingCommand::AddWiFiNetwork()
+        err = dataset.GetExtendedPanId(mExtendedPanId);
+        if (err == CHIP_NO_ERROR)
+        {
+            return ByteSpan(mExtendedPanId, sizeof(mExtendedPanId));
```
```suggestion
            return ByteSpan(mExtendedPanId);
```

bluebin14 replied:
Addressed all reviews but it got autoclosed during force push, had to open new one because Reopen button is disabled.

https://github.com/project-chip/connectedhomeip/pull/8245

bzbarsky-apple commented:
To be clear, the "changes requested" is largely for @vivien-apple's comments to get addressed.

bluebin14 commented:
Addressed all reviews but it got autoclosed during force push, had to open new one because Reopen button is disabled.

https://github.com/project-chip/connectedhomeip/pull/8111
